graphql_compiler/macros/macro_expansion.py

```diff
@@ -47,75 +67,105 @@ def _merge_selection_sets(selection_set_a, selection_set_b):
+
+    Mutate the target_ast, merging into it everything from the selection_ast. If the target
+    is at a type coercion and the selection_ast starts with a type coercion, combine them
+    into one coercion that preserves the semantics of nested coercions, which are disallowed.
+
+    For details on how fields and directives are merged, see _merge_selection_sets.
+
+    Args:
+        target_ast: AST at the @macro_edge_target directive
+        selection_ast: AST to merge inside the target. Required to have a nonempty selection set.
+        subclass_sets: optional dict mapping class names to the set of its subclass names
+    """
+    if selection_ast.selection_set is None or not selection_ast.selection_set.selections:
+        raise AssertionError(u'Precondition violated. selection_ast is expected to be nonempty {}'
+                             .format(selection_ast))
+
+    # See if there's a type coercion in the selection_ast
+    coercion = None
+    for selection in selection_ast.selection_set.selections:
+        if isinstance(selection, InlineFragment):
+            if len(selection_ast.selection_set.selections) != 1:
+                raise GraphQLCompilationError(u'Found selections outside type coercion {}'
+                                              .format(selection_ast))
+            else:
+                coercion = selection
+
     # Deal with coercions at the target
     continuation_ast = selection_ast
-    first_selection = selection_ast.selection_set.selections[0]
-    if isinstance(first_selection, InlineFragment):
-        coercion_class = first_selection.type_condition.name.value
+    if coercion is not None:
+        coercion_class = coercion.type_condition.name.value
         if isinstance(target_ast, InlineFragment):
             target_class = target_ast.type_condition.name.value
-            continuation_ast = first_selection
-            target_ast.type_condition = first_selection.type_condition
+            continuation_ast = coercion
+            target_ast.type_condition = coercion.type_condition
 
             # coercion_class is required to subclass target_class so we can merge the type coercions
             if coercion_class != target_class:
                 if subclass_sets is None:
-                    raise AssertionError(u'Cannot prove type coercion at macro target is valid. '
-                                         u'Please provide a proof that {} subclasses {} using the '
-                                         u'subclass_sets argument.'
-                                         .format(coercion_class, target_class))
+                    # TODO(bojanserafimov): Write test for this failure
+                    raise GraphQLCompilationError(
+                        u'Cannot prove type coercion at macro target is valid. Please provide a'
+                        u'proof that {} subclasses {} using the subclass_sets argument.'
```

Comment:
Not sure that `proof` is the word we're looking for here... I think the user could conceivably "prove" something that is actually false.